### PR TITLE
Improve Offline Chat Model Experience

### LINKED ISSUE #406: Invalid model file (llama v2)
Continuing discussion from #379 with @authorfunction 

I have previously gotten search to work in isolation (without chat activated) and vice versa (chat wokring before activating search); pasting the latest error message/log below. Looks to me to be identical to what I pasted earlier in this thread. Khoj shut down so I need to restart/reindex and deactivate chat to be able to confirm that search is working like previously, will take some time.

```python
           INFO     127.0.0.1:59787 - "POST                                                                                 h11_impl.py:431
                    /api/config/data/processor/conversation/enable_offline_chat?enable_offline_chat=false HTTP/1.1" 200
[11:05:02] INFO     💬 Setting up conversation processor                                                                   configure.py:286
Found model file at  /Users/mattslindstrom/.cache/gpt4all/llama-2-7b-chat.ggmlv3.q4_K_S.bin
Invalid model file
           INFO     127.0.0.1:59787 - "POST                                                                                 h11_impl.py:431
                    /api/config/data/processor/conversation/enable_offline_chat?enable_offline_chat=true HTTP/1.1" 500
           ERROR    Exception in ASGI application                                                                           h11_impl.py:369

                    ╭───────────────────────────────── Traceback (most recent call last) ─────────────────────────────────╮
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/uvicorn/protocols/http/h │
                    │ 11_impl.py:366 in run_asgi                                                                          │
                    │                                                                                                     │
                    │   363 │   # ASGI exception wrapper                                                                  │
                    │   364 │   async def run_asgi(self, app):                                                            │
                    │   365 │   │   try:                                                                                  │
                    │ ❱ 366 │   │   │   result = await app(self.scope, self.receive, self.send)                           │
                    │   367 │   │   except BaseException as exc:                                                          │
                    │   368 │   │   │   msg = "Exception in ASGI application\n"                                           │
                    │   369 │   │   │   self.logger.error(msg, exc_info=exc)                                              │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/uvicorn/middleware/proxy │
                    │ _headers.py:75 in __call__                                                                          │
                    │                                                                                                     │
                    │   72 │   │   │   │   │   port = 0                                                                   │
                    │   73 │   │   │   │   │   scope["client"] = (host, port)  # type: ignore[index]                      │
                    │   74 │   │                                                                                          │
                    │ ❱ 75 │   │   return await self.app(scope, receive, send)                                            │
                    │   76                                                                                                │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/fastapi/applications.py: │
                    │ 269 in __call__                                                                                     │
                    │                                                                                                     │
                    │   266 │   async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:             │
                    │   267 │   │   if self.root_path:                                                                    │
                    │   268 │   │   │   scope["root_path"] = self.root_path                                               │
                    │ ❱ 269 │   │   await super().__call__(scope, receive, send)                                          │
                    │   270 │                                                                                             │
                    │   271 │   def add_api_route(                                                                        │
                    │   272 │   │   self,                                                                                 │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/applications.p │
                    │ y:124 in __call__                                                                                   │
                    │                                                                                                     │
                    │   121 │                                                                                             │
                    │   122 │   async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:             │
                    │   123 │   │   scope["app"] = self                                                                   │
                    │ ❱ 124 │   │   await self.middleware_stack(scope, receive, send)                                     │
                    │   125 │                                                                                             │
                    │   126 │   # The following usages are now discouraged in favour of configuration                     │
                    │   127 │   # during Starlette.__init__(...)                                                          │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/middleware/err │
                    │ ors.py:184 in __call__                                                                              │
                    │                                                                                                     │
                    │   181 │   │   │   # We always continue to raise the exception.                                      │
                    │   182 │   │   │   # This allows servers to log the error, or allows test clients                    │
                    │   183 │   │   │   # to optionally raise the error within the test case.                             │
                    │ ❱ 184 │   │   │   raise exc                                                                         │
                    │   185 │                                                                                             │
                    │   186 │   def format_line(                                                                          │
                    │   187 │   │   self, index: int, line: str, frame_lineno: int, frame_index: int                      │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/middleware/err │
                    │ ors.py:162 in __call__                                                                              │
                    │                                                                                                     │
                    │   159 │   │   │   await send(message)                                                               │
                    │   160 │   │                                                                                         │
                    │   161 │   │   try:                                                                                  │
                    │ ❱ 162 │   │   │   await self.app(scope, receive, _send)                                             │
                    │   163 │   │   except Exception as exc:                                                              │
                    │   164 │   │   │   request = Request(scope)                                                          │
                    │   165 │   │   │   if self.debug:                                                                    │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/exceptions.py: │
                    │ 93 in __call__                                                                                      │
                    │                                                                                                     │
                    │    90 │   │   │   │   handler = self._lookup_exception_handler(exc)                                 │
                    │    91 │   │   │                                                                                     │
                    │    92 │   │   │   if handler is None:                                                               │
                    │ ❱  93 │   │   │   │   raise exc                                                                     │
                    │    94 │   │   │                                                                                     │
                    │    95 │   │   │   if response_started:                                                              │
                    │    96 │   │   │   │   msg = "Caught handled exception, but response already started."               │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/exceptions.py: │
                    │ 82 in __call__                                                                                      │
                    │                                                                                                     │
                    │    79 │   │   │   await send(message)                                                               │
                    │    80 │   │                                                                                         │
                    │    81 │   │   try:                                                                                  │
                    │ ❱  82 │   │   │   await self.app(scope, receive, sender)                                            │
                    │    83 │   │   except Exception as exc:                                                              │
                    │    84 │   │   │   handler = None                                                                    │
                    │    85                                                                                               │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/fastapi/middleware/async │
                    │ exitstack.py:21 in __call__                                                                         │
                    │                                                                                                     │
                    │   18 │   │   │   │   │   await self.app(scope, receive, send)                                       │
                    │   19 │   │   │   │   except Exception as e:                                                         │
                    │   20 │   │   │   │   │   dependency_exception = e                                                   │
                    │ ❱ 21 │   │   │   │   │   raise e                                                                    │
                    │   22 │   │   │   if dependency_exception:                                                           │
                    │   23 │   │   │   │   # This exception was possibly handled by the dependency but it should          │
                    │   24 │   │   │   │   # still bubble up so that the ServerErrorMiddleware can return a 500           │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/fastapi/middleware/async │
                    │ exitstack.py:18 in __call__                                                                         │
                    │                                                                                                     │
                    │   15 │   │   │   async with AsyncExitStack() as stack:                                              │
                    │   16 │   │   │   │   scope[self.context_name] = stack                                               │
                    │   17 │   │   │   │   try:                                                                           │
                    │ ❱ 18 │   │   │   │   │   await self.app(scope, receive, send)                                       │
                    │   19 │   │   │   │   except Exception as e:                                                         │
                    │   20 │   │   │   │   │   dependency_exception = e                                                   │
                    │   21 │   │   │   │   │   raise e                                                                    │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/routing.py:670 │
                    │ in __call__                                                                                         │
                    │                                                                                                     │
                    │   667 │   │   │   match, child_scope = route.matches(scope)                                         │
                    │   668 │   │   │   if match == Match.FULL:                                                           │
                    │   669 │   │   │   │   scope.update(child_scope)                                                     │
                    │ ❱ 670 │   │   │   │   await route.handle(scope, receive, send)                                      │
                    │   671 │   │   │   │   return                                                                        │
                    │   672 │   │   │   elif match == Match.PARTIAL and partial is None:                                  │
                    │   673 │   │   │   │   partial = route                                                               │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/routing.py:266 │
                    │ in handle                                                                                           │
                    │                                                                                                     │
                    │   263 │   │   │   │   )                                                                             │
                    │   264 │   │   │   await response(scope, receive, send)                                              │
                    │   265 │   │   else:                                                                                 │
                    │ ❱ 266 │   │   │   await self.app(scope, receive, send)                                              │
                    │   267 │                                                                                             │
                    │   268 │   def __eq__(self, other: typing.Any) -> bool:                                              │
                    │   269 │   │   return (                                                                              │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/starlette/routing.py:65  │
                    │ in app                                                                                              │
                    │                                                                                                     │
                    │    62 │   async def app(scope: Scope, receive: Receive, send: Send) -> None:                        │
                    │    63 │   │   request = Request(scope, receive=receive, send=send)                                  │
                    │    64 │   │   if is_coroutine:                                                                      │
                    │ ❱  65 │   │   │   response = await func(request)                                                    │
                    │    66 │   │   else:                                                                                 │
                    │    67 │   │   │   response = await run_in_threadpool(func, request)                                 │
                    │    68 │   │   await response(scope, receive, send)                                                  │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/fastapi/routing.py:227   │
                    │ in app                                                                                              │
                    │                                                                                                     │
                    │    224 │   │   if errors:                                                                           │
                    │    225 │   │   │   raise RequestValidationError(errors, body=body)                                  │
                    │    226 │   │   else:                                                                                │
                    │ ❱  227 │   │   │   raw_response = await run_endpoint_function(                                      │
                    │    228 │   │   │   │   dependant=dependant, values=values, is_coroutine=is_coroutine                │
                    │    229 │   │   │   )                                                                                │
                    │    230                                                                                              │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/fastapi/routing.py:160   │
                    │ in run_endpoint_function                                                                            │
                    │                                                                                                     │
                    │    157 │   assert dependant.call is not None, "dependant.call must be a function"                   │
                    │    158 │                                                                                            │
                    │    159 │   if is_coroutine:                                                                         │
                    │ ❱  160 │   │   return await dependant.call(**values)                                                │
                    │    161 │   else:                                                                                    │
                    │    162 │   │   return await run_in_threadpool(dependant.call, **values)                             │
                    │    163                                                                                              │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/khoj/routers/api.py:299  │
                    │ in set_processor_enable_offline_chat_config_data                                                    │
                    │                                                                                                     │
                    │   296 │   │                                                                                         │
                    │   297 │   │   assert state.config.processor.conversation is not None                                │
                    │   298 │   │   state.config.processor.conversation.enable_offline_chat = enable_offline_chat         │
                    │ ❱ 299 │   │   state.processor_config = configure_processor(state.config.processor,                  │
                    │       state.processor_config)                                                                       │
                    │   300 │   │                                                                                         │
                    │   301 │   │   update_telemetry_state(                                                               │
                    │   302 │   │   │   request=request,                                                                  │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/khoj/configure.py:287 in │
                    │ configure_processor                                                                                 │
                    │                                                                                                     │
                    │   284 │                                                                                             │
                    │   285 │   # Initialize Conversation Processor                                                       │
                    │   286 │   logger.info("💬 Setting up conversation processor")                                       │
                    │ ❱ 287 │   processor.conversation = configure_conversation_processor(processor_config,               │
                    │       state_processor_config)                                                                       │
                    │   288 │                                                                                             │
                    │   289 │   return processor                                                                          │
                    │   290                                                                                               │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/khoj/configure.py:316 in │
                    │ configure_conversation_processor                                                                    │
                    │                                                                                                     │
                    │   313 │   │   │   )                                                                                 │
                    │   314 │   │   )                                                                                     │
                    │   315 │   else:                                                                                     │
                    │ ❱ 316 │   │   conversation_processor = ConversationProcessorConfigModel(                            │
                    │   317 │   │   │   conversation_config=processor_config.conversation,                                │
                    │   318 │   │   )                                                                                     │
                    │   319 │   │   conversation_logfile =                                                                │
                    │       resolve_absolute_path(conversation_processor.conversation_logfile)                            │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/khoj/utils/config.py:98  │
                    │ in __init__                                                                                         │
                    │                                                                                                     │
                    │    95 │   │   self.meta_log: dict = {}                                                              │
                    │    96 │   │                                                                                         │
                    │    97 │   │   if self.enable_offline_chat:                                                          │
                    │ ❱  98 │   │   │   self.gpt4all_model.loaded_model =                                                 │
                    │       download_model(self.gpt4all_model.chat_model)                                                 │
                    │    99 │   │   else:                                                                                 │
                    │   100 │   │   │   self.gpt4all_model.loaded_model = None                                            │
                    │   101                                                                                               │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/khoj/processor/conversat │
                    │ ion/gpt4all/utils.py:20 in download_model                                                           │
                    │                                                                                                     │
                    │   17 │                                                                                              │
                    │   18 │   filename = os.path.expanduser(f"~/.cache/gpt4all/{model_name}")                            │
                    │   19 │   if os.path.exists(filename):                                                               │
                    │ ❱ 20 │   │   return GPT4All(model_name)                                                             │
                    │   21 │                                                                                              │
                    │   22 │   try:                                                                                       │
                    │   23 │   │   os.makedirs(os.path.dirname(filename), exist_ok=True)                                  │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/gpt4all/gpt4all.py:90 in │
                    │ __init__                                                                                            │
                    │                                                                                                     │
                    │    87 │   │   self.config: ConfigType = self.retrieve_model(                                        │
                    │    88 │   │   │   model_name, model_path=model_path, allow_download=allow_download                  │
                    │    89 │   │   )                                                                                     │
                    │ ❱  90 │   │   self.model.load_model(self.config["path"])                                            │
                    │    91 │   │   # Set n_threads                                                                       │
                    │    92 │   │   if n_threads is not None:                                                             │
                    │    93 │   │   │   self.model.set_thread_count(n_threads)                                            │
                    │                                                                                                     │
                    │ /Users/mattslindstrom/opt/anaconda3/envs/khoj/lib/python3.10/site-packages/gpt4all/pyllmodel.py:192 │
                    │ in load_model                                                                                       │
                    │                                                                                                     │
                    │   189 │   │   if self.model is not None:                                                            │
                    │   190 │   │   │   llmodel.llmodel_loadModel(self.model, model_path_enc)                             │
                    │   191 │   │   else:                                                                                 │
                    │ ❱ 192 │   │   │   raise ValueError("Unable to instantiate model")                                   │
                    │   193 │   │                                                                                         │
                    │   194 │   │   filename = os.path.basename(model_path)                                               │
                    │   195 │   │   self.model_name = os.path.splitext(filename)[0]                                       │
                    ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
                    ValueError: Unable to instantiate model
[11:07:34] WARNING  Unable to process file:                                                                             pdf_to_jsonl.py:101
                    /Users/mattslindstrom/GoogleDrive/LITTERATURARKIV/böcker_texter_artiklar/Digital humaniora,
                    texter/sem1/Rosenberg Early Modern Information Overload.pdf. This file will not be indexed.
           WARNING  Odd-length string                                                                                   pdf_to_jsonl.py:102
[11:09:18] WARNING  Unable to process file:                                                                             pdf_to_jsonl.py:101
                    /Users/mattslindstrom/GoogleDrive/LITTERATURARKIV/böcker_texter_artiklar/Gelb IJ A study of writing
                    a grammatology 1952.pdf. This file will not be indexed.
           WARNING  PyCryptodome is required for AES algorithm                                                          pdf_to_jsonl.py:102
[11:39:42] WARNING  Unable to process file:                                                                             pdf_to_jsonl.py:101
                    /Users/mattslindstrom/GoogleDrive/LITTERATURARKIV/böcker_texter_artiklar/Rosenberg Early Modern
                    Information Overload.pdf. This file will not be indexed.
           WARNING  Odd-length string                                                                                   pdf_to_jsonl.py:102
[12:09:25] WARNING   impossible to decode XFormObject /Bg                                                                     _utils.py:450
^C[12:29:48] INFO     Shutting down                                                                                             server.py:252
           INFO     Waiting for application shutdown.                                                                              on.py:64
           INFO     Application shutdown complete.                                                                                 on.py:75
           INFO     Finished server process [54595]                                                                            server.py:85
           INFO     🌒 Stopping Khoj                                                                                            main.py:160
```

### PR DESCRIPTION
**Pull request by debanjum**
- a85ff94 Make offline chat model user configurable. Use `filename` of any [GPT4All supported  model](https://github.com/nomic-ai/gpt4all/blob/main/gpt4all-chat/metadata/models.json) like below:

  ```yml
  ...
  processor:
    conversation:
      enable-offline-chat: true
      offline-chat-model: wizardlm-13b-v1.1-superhot-8k.ggmlv3.q4_0.bin
  ...
  ```
- d1ff812 Run GPT4All Chat Model on GPU, when available via [GPT4All Vulcan support](https://blog.nomic.ai/posts/gpt4all-gpu-inference-with-vulkan)
- 13b16a4 Use default Llama 2 supported by GPT4All
- Make `tokenizer` and `max-prompt-size` of chat model user configurable. E.g When using chat models not in [this pre-defined list](https://github.com/khoj-ai/khoj/blob/master/src/khoj/processor/conversation/utils.py) that support larger context window or a different tokenizer.

Closes #406

### CODE REVIEW

File: src/khoj/processor/conversation/prompts.py [hunk 1]
```diff
@@ -91,7 +91,7 @@
 
 notes_conversation_llamav2 = PromptTemplate.from_template(
     """
-Notes:
+Users Notes:
```
Comment (sabaimran) [inline src/khoj/processor/conversation/prompts.py:94]:
```suggestion
User's Notes:
```

**Comment by sabaimran** [comment]
Great to be using built-in support for Llama V2 via GPT4All going forward! A little confused about the PR description. You said "Make offline chat model user configurable", but it's still limited to only Llama V2, unless I'm missing something. What did you mean by that?

**Review by sabaimran — APPROVED** [review]

File: src/khoj/processor/conversation/prompts.py (hunk 1 above)
Comment (debanjum) [reply src/khoj/processor/conversation/prompts.py]: Fixed!

**Comment by debanjum** [comment]
> You said "Make offline chat model user configurable", but it's still limited to only Llama V2, unless I'm missing something. What did you mean by that?

Updated the PR description with more details on how to use a different offline chat model. Made [a few fixes](https://github.com/khoj-ai/khoj/pull/494/commits/1ad8b150e88061d5cea295b610be2185c8532047) to use default fallback max_prompt_size and tokenizer to make this work

**Pull request merged**